SYS-1517: Create basic app

ztucker4 opened the pull request:
Implements [SYS-1517](https://uclalibrary.atlassian.net/browse/SYS-1517)

This PR implements the basic functionality of the digital signage hours app. Including:

* One model for Location, containing a name to use on the sign and the LibCal location ID.
  * Sign orientation options (landscape, small portrait, and large portrait) are not implemented as models, but as form choices eventually used by the display view.
* A view (`display_hours`) that gets hours information from the LibCal Hours widget and formats them, and a corresponding template to display the hours on signs of different sizes. All existing locations work with all sign orientations.
  * These displays are intended to match [the legacy examples](https://github.com/UCLALibrary/digital-signage-hours/tree/fd3f535e2d861e3dc078f21dd76ac325f385b4cd/sample_html). 
* A view and template to allow users to generate URLs for specified location and sign orientations.
* Views and templates for displaying logs and release notes (reused from previous Django projects).

To test, apply migrations and load the locations fixture, then go to http://127.0.0.1:8000/. Other than hours display, all parts of the application are accessible from the navbar.

Limitations
* Automated testing is currently very minimal.
* This PR doesn't include anything related to the multicolumn events schedule display, which I split into a new ticket/branch (SYS-1524).

[SYS-1517]: https://uclalibrary.atlassian.net/browse/SYS-1517?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

akohler commented on signs/views_utils.py:
```diff
@@ -0,0 +1,56 @@
+import requests
+import logging
+from datetime import datetime
+from bs4 import BeautifulSoup
```
@ztucker4 This package is not defined in `requirements.txt`, so the application crashes on startup.  I don't see any actual uses of this package.

For now I've commented it out locally so I can continue reviewing, but please fix this.

akohler commented on signs/views.py:
```diff
@@ -1,3 +1,87 @@
+                    "url": url,
+                },
+            )
+    return render(
```
These 2 `return` statements can be combined.  Just make sure `url` is initialized as `None` before the `return`.

Since there are several variables being passed, use a `context`, similar to what's in `display_hours`.

akohler commented on signs/templates/signs/get_URL.html:
```diff
@@ -0,0 +1,16 @@
+    </form>
+    {% if url %}
+        <p>Paste this URL into Rise Vision:</p>
+        <p>{{ request.get_host }}{{ url }}</p>
```
Make the URL clickable with [`urlize`](https://docs.djangoproject.com/en/4.2/ref/templates/builtins/#urlize) for convenience?

akohler commented on signs/views.py:
```diff
@@ -1,3 +1,87 @@
+) -> HttpResponse:
+    """Display hours for a location. This view is used by the digital signage system."""
+
+    widget_url = "https://calendar.library.ucla.edu/widget/hours/grid?"
```
Let's move this to `settings.py`, via a variable defined in `.docker-compose_django.env` (and later, in the Helm chart).  Then if this URL changes, the program code doesn't need to be updated, just the variable value.

Once in `settings.py`, add an import:
`from django.conf import settings`

Then the value can be referenced as `settings.WIDGET_URL` (or whatever).

akohler commented on signs/views.py:
```diff
@@ -1,3 +1,87 @@
+
+
+@login_required
+def show_log(request, line_count: int = 200) -> HttpResponse:
```
Please add a type hint for `request` parameter.

akohler reviewed (changes requested):
@ztucker4 Overall, this is an excellent start.  The application works (once the bad import of `bs4` is removed) and the output looks great to me.

Most of these comments are small improvements, intended for consistency / readability.  We can talk about any of them at standup / team meeting.

Thanks --Andy

akohler commented on requirements.txt:
```diff
@@ -2,3 +2,5 @@ Django==4.2.7
 whitenoise==6.5.0
 gunicorn==21.2.0
+requests==2.31.0
+django-bootstrap5 == 23.3
```
Please follow other format, no spaces around `==`.

akohler commented on signs/templates/signs/base.html:
```diff
@@ -0,0 +1,32 @@
+
+<body>
+    <ul class="navbar">
+        <li><a href="/get_hours_URL/">Generate Hours URL</a></li>
```
Change `URL` to `url` to match view; this currently gives 404 when clicked.

akohler commented on signs/views.py:
```diff
@@ -1,3 +1,81 @@
+    get_start_end_dates,
+    construct_display_url,
+)
+from signs.forms import LocationForm  # , DisplayTypeForm
```
Remove comment since this form no longer exists.

akohler commented on signs/views_utils.py:
```diff
@@ -0,0 +1,78 @@
+
+    # Days dicts contain data we don't need, so reformat into a new list of dicts
+    hours = []
+    for day in days:
```
Unfortunately, this no longer works as expected.  The raw data in `weeks[0]` is ordered `Sunday..Saturday`.  Replacing Sunday's data winds up with the later Sunday being the first day.  Stepping through the code in the shell:
```
>>> from signs.views_utils import *
>>> from django.conf import settings
>>> from pprint import pprint

>>> widget_url = settings.LIBCAL_HOURS_WIDGET
>>> libcal_id = 4690 # Arts
>>> hours = get_hours(widget_url, libcal_id)
>>> f_hours = format_hours(hours)

>>> for d in f_hours: print(d)
...
{'date': '2024-02-11', 'weekday': 'Sunday', 'rendered_hours': 'Closed'}
{'date': '2024-02-05', 'weekday': 'Monday', 'rendered_hours': '9am - 9pm'}
{'date': '2024-02-06', 'weekday': 'Tuesday', 'rendered_hours': '9am - 9pm'}
{'date': '2024-02-07', 'weekday': 'Wednesday', 'rendered_hours': '9am - 9pm'}
{'date': '2024-02-08', 'weekday': 'Thursday', 'rendered_hours': '9am - 9pm'}
{'date': '2024-02-09', 'weekday': 'Friday', 'rendered_hours': '9am - 5pm'}
{'date': '2024-02-10', 'weekday': 'Saturday', 'rendered_hours': '1pm - 5pm'}
```

See for example http://127.0.0.1:8000/display_hours/4690/landscape, which for me currently looks like this:
![signs_misordered_days](https://github.com/UCLALibrary/digital-signage-hours/assets/2213836/c1a64ef3-6ebb-441c-8c5b-b490dade5958)

akohler commented on signs/views_utils.py:
```diff
@@ -0,0 +1,78 @@
+    month-day format, e.g. ("Feb 05","Feb 11")."""
+    # hours list is constructed in order, so we can just take the first and last items
+    start = hours[0]["date"]
+    start = datetime.strptime(start, "%Y-%m-%d")
```
Pull repeated code (the `strptime` and `strftime`) into a helper method, `format_date(date)` or similar.  This makes it easier to change the format if needed, and to test the formatting of the dates in isolation.

akohler reviewed (changes requested):
@ztucker4 Almost everything has been addressed, but the refactoring of `format_hours()` introduced a sequencing bug.  Please add a test for that immediately, to help ensure it's correct.

Thanks --Andy

ztucker4 commented:
Hi @akohler , this is ready for yet another look. I fixed the issue with date ranges and added tests for that and related functions.

I also tried to add some safeguards in` format_hours()` (in `views_utils.py`) so that we can display more user-friendly error messages if something goes wrong in getting LibCal data. You can see this if you add a fake location with a nonexistent ID in the Admin interface and then try to display hours for it.

akohler commented on signs/views_utils.py:
```diff
@@ -0,0 +1,101 @@
+
+    # First, check that the data contains the expected number of values:
+    #  1 location, 2 weeks, 7 days each
+    if len(data) != 1:
```
There are still some problems.  For example, Arts Library http://127.0.0.1:8000/display_hours/4690/landscape triggers this error, because `len(data) == 2`:
`dict_keys(['loc_4690', 'loc_20525'])`
That's Arts Library itself (4690) and Arts Library Reference Desk (20525).

Since signs should show only hours for one location, probably `get_hours()` should filter the response to include just the data for the specified location.

akohler replied:
I checked the size of `data` for each location, and several have multiple locs in them, as many as 6:
```
19256 1
4690 2
20525 1
2081 3
2607 1
2609 1
2608 1
2614 1
21770 1
4693 1
3291 1
4705 1
4706 1
2082 1
22368 2
4694 2
4695 1
6133 1
2083 1
22369 1
4696 2
22392 1
4699 1
2572 6
20241 1
1916 4
3280 1
4702 4
4703 1
20242 1
```

akohler replied:
Confirmed `get_hours()` now returns only one location's data for each location currently in the database.

akohler reviewed (changes requested):
@ztucker4 The added tests and messages are good.  Users may want a different message to appear on the sign if there's an error, but we'll address that later, if needed.

I think the only remaining problem is incorrect handling of locations where the widget returns data for multiple locs, like Arts Library (http://127.0.0.1:8000/display_hours/4690/landscape).  Since the signs apparently are intended to show hours for a single location only, I think the fix is to filter the widget's response to return only the data for the requested location, which should be a simple change.

ztucker4 commented:
I made the changes we discussed at standup. Extracting the correct location from the LibCal response is now done in `get_hours()`, and the error cases in `format_hours()` have been cleaned up.

akohler commented:
@ztucker4 The changes all look good.  Tests pass and make sense.  Merging this!
